src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -1178,6 +1178,30 @@ protected function serialize_token(): string {
 		$token_type = $this->get_token_type();
 
 		switch ( $token_type ) {
+			case '#doctype':
+				$doctype = $this->get_doctype_info();
+				if ( null === $doctype ) {
+					break;
+				}
+
+				$html .= '<!DOCTYPE';
+
+				if ( $doctype->name ) {
+					$html .= " {$doctype->name}";
+				}
+
+				if ( null !== $doctype->public_identifier ) {
+					$html .= " PUBLIC \"{$doctype->public_identifier}\"";
+				}
+				if ( null !== $doctype->system_identifier ) {
+					if ( null === $doctype->public_identifier ) {
+						$html .= ' SYSTEM';
+					}
+					$html .= " \"{$doctype->system_identifier}\"";
+				}
+				$html .= '>';
```

Comment:
The output would be nicer if there were a newline here. It's not necessary but would look better. I'm pretty sure that at this position it's completely invisible to parsers.

Curious to hear thoughts 🤔